[GTK] Crash in GraphicsContextGLGBM::allocateDrawBufferObject

carlosgcampos opened the pull request:
#### 153153309cef73f2b65f8dfc3991044e705fe5a5
<pre>
[GTK] Crash in GraphicsContextGLGBM::allocateDrawBufferObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=255398">https://bugs.webkit.org/show_bug.cgi?id=255398</a>

Reviewed by Michael Catanzaro.

GBMBufferSwapchain::getBuffer() can return nullptr for several reasons,
so we need to handle that case in
GraphicsContextGLGBM::allocateDrawBufferObject(). This also adds error
messages for the cases where getBuffer() returns nullptr, and returns
earlier without creating any buffer when the given format is
unsupported.

* Source/WebCore/platform/graphics/gbm/GBMBufferSwapchain.cpp:
(WebCore::isBufferFormatSupported):
(WebCore::GBMBufferSwapchain::getBuffer):
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.cpp:
(WebCore::GraphicsContextGLGBM::allocateDrawBufferObject):

Canonical link: <a href="https://commits.webkit.org/264648@main">https://commits.webkit.org/264648@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f7a6eb2923fff231499c9d1d79ce31988244395b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8004 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8289 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8501 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9656 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8108 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8008 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10275 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8200 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10974 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8148 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9236 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7313 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9776 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6537 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7325 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14906 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7660 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7446 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10809 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7916 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6439 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7230 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11438 "Built successfully") | | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/978 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7657 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

zdobersek commented on Source/WebCore/platform/graphics/gbm/GBMBufferSwapchain.cpp:
```diff
@@ -43,11 +43,49 @@ GBMBufferSwapchain::GBMBufferSwapchain(BufferSwapchainSize size)
+    if (!isBufferFormatSupported(description.format)) {
+        WTFLogAlways("Failed to get GBM buffer from swap chain: unsupported format");
+        return nullptr;
+    }
```
This position can unnecessarily leave live buffer objects that would otherwise be destroyed in the next if-clause.

mcatanzaro reviewed (approved):
I think we should land this now to get it into the imminent 2.40.2 release. Zan, if this lands before you get a chance to provide further feedback, Carlos will need to follow up.

carlosgcampos committed:
[GTK] Crash in GraphicsContextGLGBM::allocateDrawBufferObject
https://bugs.webkit.org/show_bug.cgi?id=255398

Reviewed by Michael Catanzaro.

GBMBufferSwapchain::getBuffer() can return nullptr for several reasons,
so we need to handle that case in
GraphicsContextGLGBM::allocateDrawBufferObject(). This also adds error
messages for the cases where getBuffer() returns nullptr, and returns
earlier without creating any buffer when the given format is
unsupported.

* Source/WebCore/platform/graphics/gbm/GBMBufferSwapchain.cpp:
(WebCore::isBufferFormatSupported):
(WebCore::GBMBufferSwapchain::getBuffer):
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.cpp:
(WebCore::GraphicsContextGLGBM::allocateDrawBufferObject):

Canonical link: https://commits.webkit.org/264648@main
